### ISSUE TITLE
Reversing chromaprints back into chromagrams

### ISSUE THREAD

**Issue by Cortexelus** [acoustid/chromaprint#123]
Hi
I'm interested in reversing chromaprints back into their chromagrams (a rough estimation)
The https://oxygene.sk/2011/01/how-does-chromaprint-work/ explanation is a great starting point, and I think I could figure out what the filters were from studying the source code.
But just checking, I'm guessing there's probably a python gist for this somewhere already
Thanks, 
CJ

**Comment by Cortexelus** [comment]
Update: After re-implementing your fingerprinter's classifier in python, I realize to what extent the reversal is non-deterministic. 

*Dequantizer*: assuming the 3 points are the 25% 50% and 75% medians of a skewed gaussian, I can sample from it to get plausible filter magnitudes. 
*Defiltering*: but the filters are complex to detangle. we know their magnitudes but not their centers 

Seems like an optimization problem. I could re-implement the fingerprinter differentiably in pytorch, then use gradient descent to solve it.

**Comment by lalinsky** [comment]
@Cortexelus You are sure this is worth the effort? They are not built for that. The chroma information at the source is really junk. I didn't know what I was doing back then. It's all just hacked together. It works well for the purpose for which it was built, but there is no musically meaningful information in the fingerprint before the hashing happens. What's your goal with this?

**Comment by Cortexelus** [comment]
My goal is fingerprint-conditioned neural synthesis -- I'm curious see what happens if I use a fingerprint to hallucinate the rest of the audio (in 48k stereo). There are many possible solutions for each fingerprint, I think it'd be way interesting to hear all the variations! I understand the format is minimal, but I bet I knows just enough about song structure and repetition to be interesting as the initial conditioning signal (before it gets progressively upsampled into audio).

Responding to my previous comment: I concluded there's no point in having a chromagram solver middleman. I mistakenly thought in order for a generator to run real-time I needed to disentangle the filters' time-windows, but actually it's nice that each fingerprint knows something about the future context. I'm just going to use the raw (de-graycoded) filter weights as the conditioning signal w/ attention & leave it up to the net to figure out what's meaningful. 

Acoustid is the largest open library of fingerprints that I know of, in the tens of millions, and the shear size of that makes it cool to work with.
